app/src/main/java/com/p2p/presentation/basegame/GameActivity.kt

```diff
@@ -0,0 +1,136 @@
+import android.content.Intent
+import android.os.Handler
+import android.os.Looper
+import androidx.annotation.LayoutRes
+import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
+import com.google.android.material.dialog.MaterialAlertDialogBuilder
+import com.p2p.R
+import com.p2p.framework.bluetooth.BluetoothConnectionThread.Companion.MESSAGE_READ
+import com.p2p.framework.bluetooth.BluetoothConnectionThread.Companion.MESSAGE_WRITE_ERROR
+import com.p2p.framework.bluetooth.BluetoothConnectionThread.Companion.MESSAGE_WRITE_SUCCESS
+import com.p2p.framework.bluetooth.BluetoothConnectionThread.Companion.SENDER_ID
+import com.p2p.model.message.Message
+import com.p2p.model.message.MessageReceived
+import com.p2p.presentation.base.BaseMVVMActivity
+import com.p2p.utils.Logger
+import kotlin.reflect.KClass
+
+abstract class GameActivity<E : SpecificGameEvent, VM : GameViewModel>(
+    @LayoutRes layout: Int = R.layout.activity_base
+) : BaseMVVMActivity<GameEvent, VM>(layout) {
+
+    protected val gameViewModelFactory: GameViewModelFactory
+        get() = GameViewModelFactory(baseContext, gameViewModelFactoryData)
+
+    protected val gameViewModelFactoryData: GameViewModelFactory.Data
+        get() = GameViewModelFactory.Data(
+            handler = handler,
+            gameConnectionType = gameConnectionType,
+            device = device
+        )
+
+    private val handler = Handler(Looper.getMainLooper()) {
+        when (it.what) {
+            MESSAGE_READ -> {
+                val message = it.toMessage()
+                Logger.d(TAG, "Read: $message")
+                viewModel.receiveMessage(MessageReceived(message, it.data.getLong(SENDER_ID)))
+                true
+            }
+            MESSAGE_WRITE_SUCCESS -> {
+                val message = it.toMessage()
+                Logger.d(TAG, "Sent successfully: $message")
+                viewModel.onSentSuccessfully(message)
+                true
+            }
+            MESSAGE_WRITE_ERROR -> {
+                val message = it.toMessage()
+                Logger.d(TAG, "Cannot write: $message")
+                viewModel.onSentError(message)
+                true
+            }
+            else -> false
+        }
+    }
+    private val gameConnectionType: String by lazy {
+        requireNotNull(intent.getStringExtra(GameConnectionType.EXTRA)) {
+            "The connection type wasn't sent. Use the start method of [GameActivity]."
+        }
+    }
+    private val device: BluetoothDevice? by lazy { intent.getParcelableExtra(SERVER_DEVICE_EXTRA) }
+    private val objectMapper by lazy { jacksonObjectMapper() }
+
+    final override fun onEvent(event: GameEvent) = when (event) {
+        GoToCreate -> goToCreate()
+        GoToClientLobby -> goToClientLobby()
+        GoToServerLobby -> goToServerLobby()
+        GoToPlay -> goToPlay()
+        is OpenInstructions -> showInstructions(event.instructions)
+        is SpecificGameEvent -> {
```

Comment:
Por qué estas dos opciones necesitan el `is` y el resto no?

Reply:
`is` para cuando es una clase, no hace falta cuando es un object.